File: zone-server/src/main/kotlin/net/bestia/zone/economy/shop/ShopOfferSMSG.kt
package net.bestia.zone.economy.shop

import net.bestia.bnet.proto.EnvelopeProto
import net.bestia.bnet.proto.ShopOfferSmsgProto
import net.bestia.zone.message.SMSG
import net.bestia.zone.util.EntityId

/**
 * One settlement's shelves as a client draws them.
 *
 * A snapshot for a window, never a contract. Every trade is settled unit by unit against the live price
 * at the moment it resolves, so a client acting on a stale window pays what things cost now - which is
 * also why this is re-sent after every trade rather than being kept in step by deltas.
 */
data class ShopOfferSMSG(
  val settlement: Int,
  val merchantEntityId: EntityId,
  val entries: List<Entry>,
) : SMSG {

  /** @param offered what the town will part with: the shelves less what the locals keep back */
  data class Entry(val itemId: Long, val offered: Int, val buyPrice: Long, val sellPrice: Long)

  override fun toBnetEnvelope(): EnvelopeProto.Envelope {
    val offer = ShopOfferSmsgProto.ShopOfferSMSG.newBuilder()
      .setSettlement(settlement)
      .setMerchantEntityId(merchantEntityId)
      .addAllEntries(entries.map { entry ->
        ShopOfferSmsgProto.ShopEntry.newBuilder()
          .setItemId(entry.itemId)
          .setOffered(entry.offered)
          .setBuyPrice(entry.buyPrice)
          .setSellPrice(entry.sellPrice)
          .build()
      })
      .build()

    return EnvelopeProto.Envelope.newBuilder()
      .setShopOffer(offer)
      .build()
  }
}

File: bestia-client/src/Bnet/Message/Shop/ShopOfferSMSG.cs
using Godot;

namespace BestiaBehemothClient.Bnet.Message.Shop
{
  /// <summary>
  /// What one merchant will trade right now, and at what price.
  /// </summary>
  /// <remarks>
  /// A snapshot for drawing a window, never a contract: every trade is settled unit by unit against the
  /// live price when it resolves. Re-sent after every trade, which is how the window stays honest - so
  /// the panel re-renders wholesale on each one rather than patching the row it just changed.
  /// </remarks>
  [GlobalClass]
  public partial class ShopOfferSMSG : ISMSG
  {
    /// <summary>Dense settlement index, so one town's window can be told from another's.</summary>
    [Export] public int Settlement { get; set; }

    /// <summary>
    /// Whose counter this is. Carried rather than remembered, because the window is not always something
    /// this client asked for - asking a merchant for their wares in conversation pushes one.
    /// </summary>
    [Export] public ulong MerchantEntityId { get; set; }

    [Export] public Godot.Collections.Array<ShopEntry> Entries { get; set; } = [];

    public static ShopOfferSMSG FromProto(global::Bnet.ShopOfferSMSG proto)
    {
      var entries = new Godot.Collections.Array<ShopEntry>();
      foreach (var entry in proto.Entries)
      {
        entries.Add(ShopEntry.FromProto(entry));
      }

      return new ShopOfferSMSG
      {
        Settlement = proto.Settlement,
        MerchantEntityId = proto.MerchantEntityId,
        Entries = entries
      };
    }
  }
}
